### PR TITLE
`CaptumExplainer` forward method

### PR DESCRIPTION
**Pull request by RBendias**
Defines Explainer.__call__ for homogeneous/heterogeneous  graphs (#6375)

### CODE REVIEW

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 1]
```diff
@@ -32,6 +38,18 @@ def __init__(
         self.attribution_method = attribution_method
         self.kwargs = kwargs
 
+    def _get_mask_type(self):
+        "Based on the explainer config, return the mask type."
+        node_mask_type = self.explainer_config.node_mask_type
+        edge_mask_type = self.explainer_config.edge_mask_type
+        if node_mask_type is not None and edge_mask_type is not None:
+            mask_type = 'node_and_edge'
```
Comment (vid-koci) [inline torch_geometric/explain/algorithm/captum_explainer.py:46]: Could we introduce these three mask types as an Enum? Also in `captum.py`, possibly as a follow-up PR.

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 2]
```diff
@@ -42,8 +60,57 @@ def forward(
...
+
+        node_mask, edge_mask = convert_captum_output(
+            attributions,
+            mask_type,
+            metadata,
+        )
 
-        return Explanation(node_mask=None, edge_mask=None)
+        if hasattr(model, 'metadata'):
+            return HeteroExplanation(
```
Comment (rusty1s) [inline torch_geometric/explain/algorithm/captum_explainer.py:108]: I don't think this is correct. I think you would need to create similar to
```python
explanation = HeteroExplanation()
for node_type, node_mask in node_mask_dict.items():
    explanation[node_type].node_mask = node_mask

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 3]
```diff
@@ -42,8 +60,57 @@ def forward(
         index: Optional[Tensor] = None,
         **kwargs,
     ) -> Explanation:
+        import captum.attr  # noqa
```
Comment (rusty1s) [inline torch_geometric/explain/algorithm/captum_explainer.py:63]: Why do we need the import here?

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 4]
```diff
@@ -42,8 +60,57 @@ def forward(
...
+
+        inputs, add_forward_args = to_captum_input(
+            x,
+            edge_index,
+            mask_type,
+            *kwargs.values(),
+        )
+
+        if hasattr(model, 'metadata'):
+            metadata = model.metadata
```
Comment (rusty1s) [inline torch_geometric/explain/algorithm/captum_explainer.py:75]: I think it is safer to use `x.keys()` and `edge_index.keys()`

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 5]
```diff
@@ -42,8 +60,57 @@ def forward(
...
+            metadata = None
+            captum_model = CaptumModel(model, mask_type, index)
+
+        if isinstance(self.attribution_method, str):
+            attribution_method = getattr(
+                captum.attr,
+                self.attribution_method,
+            )(captum_model)
+        else:
+            attribution_method = self.attribution_method(captum_model)
```
Comment (rusty1s) [inline torch_geometric/explain/algorithm/captum_explainer.py:92]: Can we initialize that in `__init__`?

File: torch_geometric/explain/algorithm/captum_explainer.py (hunk 4 above)
Comment (wsad1) [reply torch_geometric/explain/algorithm/captum_explainer.py:83]: +1.

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 6]
```diff
@@ -42,8 +60,57 @@ def forward(
...
+        )
+
+        node_mask, edge_mask = convert_captum_output(
+            attributions,
+            mask_type,
+            metadata,
+        )
 
-        return Explanation(node_mask=None, edge_mask=None)
+        if hasattr(model, 'metadata'):
```
Comment (wsad1) [inline torch_geometric/explain/algorithm/captum_explainer.py]:
```suggestion
        if isinstance(x, dict):
```

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 7]
```diff
@@ -43,7 +70,51 @@ def forward(
...
+        mask_type = self._get_mask_type()
+
+        inputs, add_forward_args = to_captum_input(
+            x,
+            edge_index,
+            mask_type,
+            *kwargs.values(),
+        )
+
+        if hasattr(model, 'metadata'):
```
Comment (wsad1) [inline torch_geometric/explain/algorithm/captum_explainer.py:82]:
```suggestion
        if isinstance(x, dict):
```

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 8]
```diff
@@ -43,7 +70,51 @@ def forward(
...
+            **self.kwargs,
+        )
+
+        node_mask, edge_mask = convert_captum_output(
+            attributions,
+            mask_type,
+            metadata,
+        )
+
+        if not hasattr(model, 'metadata'):
```
Comment (wsad1) [inline torch_geometric/explain/algorithm/captum_explainer.py:109]:
```suggestion
        if not isinstance(x, dict):
```

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 9]
```diff
@@ -29,9 +35,30 @@ def __init__(
...
         self.kwargs = kwargs
 
+        if isinstance(attribution_method, str):
+            self.attribution_method = getattr(
+                captum.attr,
+                attribution_method,
+            )
+        else:
+            self.attribution_method = attribution_method
+
```
Comment (wsad1) [inline torch_geometric/explain/algorithm/captum_explainer.py:49]:
```suggestion
def supports():
    return True
```
It looks like all configs are supported. Except for probably a node mask `object`.

**Review by wsad1 — APPROVED** [review]
Looks good. 
Lets add an issue to update https://pytorch-geometric.readthedocs.io/en/latest/notes/explain.html, to show `HeteroExplanation`.

File: torch_geometric/explain/algorithm/captum_explainer.py (hunk 9 above)
Comment (RBendias) [reply torch_geometric/explain/algorithm/captum_explainer.py:49]: `common_attributes` and `object` are not supported.

File: torch_geometric/explain/algorithm/captum_explainer.py [hunk 10]
```diff
@@ -43,7 +70,51 @@ def forward(
...
+            *kwargs.values(),
+        )
+
+        if isinstance(x, dict):
+            metadata = (x.keys(), edge_index.keys())
+            captum_model = CaptumHeteroModel(
+                model,
+                mask_type,
+                index,
+                model.metadata,
```
Comment (rusty1s) [inline torch_geometric/explain/algorithm/captum_explainer.py:88]:
```suggestion
                metadata,
```

**Review by rusty1s — APPROVED** [review]

**Pull request merged**